### PR TITLE
Artifact ==> Event type

### PR DESCRIPTION
**Pull request by woksin**
Mainly change everything around artifacts to be around event types. Also updated the event types builder with additional methods and no-magic registration of event types

### CODE REVIEW

File: Samples/Basic/index.ts [hunk 1]
```diff
@@ -8,6 +8,7 @@ import { PartitionedFilterResult } from '@dolittle/sdk.events.filtering';
 import { MyEvent } from './MyEvent';
 import './MyEventHandler';
 import { Version } from '@dolittle/sdk.execution';
+import { eventTypes } from '@dolittle/sdk.protobuf';
```
Comment (woksin) [inline Samples/Basic/index.ts:11]: Found something that shouldn't be here
```suggestion
```

File: Source/artifacts/EventTypes.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+     * @param {Map<Function, Artifact>?} associations Known associations
+     */
+    constructor(private _associations: EventTypeMap<Constructor<any>> = new EventTypeMap()) {
+    }
+    /** @inheritdoc */
+    hasTypeFor(input: EventType): boolean {
+        for (const associatedArtifact of this._associations.keys()) {
+            if (this.eventTypesEquals(input, associatedArtifact)) return true;
+        }
+        return false;
```
Comment (jakhog) [inline Source/artifacts/EventTypes.ts:31]: Should be able to use `_associations.has(...)` here?

File: Source/artifacts/EventTypesFromDecorators.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+    /**
+     * Associate a type with a unique artifact identifier and optional generation.
+     * @param {Constructor} type Type to associate.
+     * @param {ArtifactId} identifier Identifier to associate with.
+     * @param {number} generation Optional generation - defaults to 0.
+     */
+    static associate(type: Constructor<any>, identifier: EventTypeId, generation?: Generation): void {
+        this.eventTypes.associate(type, new EventType(identifier, generation || Generation.first));
+    }
```
Comment (jakhog) [inline Source/artifacts/EventTypesFromDecorators.ts:24]: Is it nicer to use default value for generation here? It seems like we do that other places in the code...

File: Source/artifacts/IEventTypes.ts [hunk 4]
```diff
@@ -35,24 +35,23 @@ export interface IArtifacts {
...
 
     /**
      * Associate a type with a unique artifact identifier and optional generation.
      * @param {Constructor<any>} type Type to associate.
-     * @param {ArtifactId} identifier Identifier to associate with.
-     * @param {Generation} generation Optional generation - defaults to 0.
+     * @param {EventType} artifact Artifact to associate with.
      */
-    associate(type: Constructor<any>, identifier: ArtifactId, generation?: Generation): void;
+    associate(type: Constructor<any>, artifact: EventType): void;
```
Comment (jakhog) [inline Source/artifacts/IEventTypes.ts:56]: Rename parameter to _eventType_.

File: Source/artifacts/EventTypes.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+        if (!artifact) {
+            throw new UnableToResolveEventType(object, input);
+        }
+
+        return artifact;
+    }
+
+    /** @inheritdoc */
+    associate(type: Constructor<any>, artifact: EventType): void {
```
Comment (jakhog) [inline Source/artifacts/EventTypes.ts:81]: Rename parameter to _eventType_.

File: Source/artifacts/package.json [hunk 6]
```diff
@@ -37,8 +37,7 @@
         "@dolittle/rudiments": "5.0.1",
         "@dolittle/types": "5.0.1",
         "@dolittle/runtime.contracts": "5.1.19",
-        "is-natural-number": "^4.0.1",
-        "rxjs": "6.6.0"
+        "is-natural-number": "^4.0.1"
```
Comment (jakhog) [inline Source/artifacts/package.json:40]: In the spirit of limiting dependencies to reduce our exposure, can we get rid of this package and just implement the logic ourselves?

It is really just `Number.isSafeInteger(value) && value > 0`.

File: Source/events/EventLogSequenceNumberMustBeNaturalNumber.ts [hunk 7]
```diff
@@ -1,5 +1,6 @@
 // Copyright (c) Dolittle. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
 import { Exception } from '@dolittle/rudiments';
 
 /**
```
Comment (jakhog) [inline Source/events/EventLogSequenceNumberMustBeNaturalNumber.ts:6]: I think natural number as a concept isn't obvious for most people, a better name is probably `EventLogSequenceNumberMustBeAPositiveInteger`

**Review by jakhog — CHANGES_REQUESTED** [review]

File: Samples/Basic/MyEvent.ts [hunk 8]
```diff
@@ -1,6 +1,6 @@
 // Copyright (c) Dolittle. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
-import { eventType } from '@dolittle/sdk.events';
+import { eventType } from '@dolittle/sdk.artifacts';
```
[the remark this answers is not in the document]
Comment (woksin) [reply Samples/Basic/MyEvent.ts:3]: We could do that. That makes sense

File: Samples/Container/index.ts [hunk 9]
```diff
@@ -14,6 +14,8 @@ const client = Client
     .forMicroservice('7a6155dd-9109-4488-8f6f-c57fe4b65bfb')
     .withContainer(Container)
     .withLogging(_ => _.useWinston(_ => _.level = 'debug'))
+    .withEventTypes(eventTypes =>
+        eventTypes.register<MyEvent>(MyEvent))
```
[the remark this answers is not in the document]
Comment (woksin) [reply Samples/Container/index.ts:18]: We don't need it

File: Source/artifacts/EventTypesFromDecorators.ts (hunk 3 above)
Comment (woksin) [reply Source/artifacts/EventTypesFromDecorators.ts:24]: Yes (Y)

File: Source/artifacts/EventTypes.ts (hunk 2 above)
Comment (woksin) [reply Source/artifacts/EventTypes.ts:31]: Yep

File: Source/artifacts/package.json (hunk 6 above)
Comment (woksin) [reply Source/artifacts/package.json:40]: Done

File: Source/artifacts/EventTypes.ts (hunk 5 above)
Comment (woksin) [reply Source/artifacts/EventTypes.ts:81]: Done

File: Source/artifacts/IEventTypes.ts (hunk 4 above)
Comment (woksin) [reply Source/artifacts/IEventTypes.ts:56]: Done

File: Source/events/EventLogSequenceNumberMustBeNaturalNumber.ts (hunk 7 above)
Comment (woksin) [reply Source/events/EventLogSequenceNumberMustBeNaturalNumber.ts:6]: done

File: Source/artifacts/Generation.ts [hunk 10]
```diff
@@ -15,7 +14,7 @@ import { GenerationMustBeNaturalNumber } from './GenerationMustBeNaturalNumber';
 export class Generation extends ConceptAs<number, '@dolittle/sdk.artifacts.Generation'>{
 
     constructor(generation: number) {
-        if (!isNaturalNumber(generation, { includeZero: true })) throw new GenerationMustBeNaturalNumber();
+        if (!Number.isSafeInteger(generation) && generation < 0) throw new GenerationMustBePositiveInteger();
```
Comment (jakhog) [inline Source/artifacts/Generation.ts:17]: Almost 😛 
```suggestion
        if (!Number.isSafeInteger(generation) || generation < 0) throw new GenerationMustBePositiveInteger();
```

**Review by jakhog — CHANGES_REQUESTED** [review]
One little boolean operator left 😄

**Review by jakhog — APPROVED** [review]

**Pull request merged**